lib/tasks/result_monitor.rake

```diff
@@ -32,6 +32,51 @@ class MonitorPipelineResults
         LogUtil.log_backtrace(exception)
       end
     end
+
+    # "stalled uploads" are not pipeline jobs, but they fit in here better than
+    # anywhere else.
+    begin
+      MonitorPipelineResults.alert_stalled_uploads!
+    rescue => exception
+      LogUtil.log_err_and_airbrake("Failed to alert on stalled uploads: #{exception.message}")
+      LogUtil.log_backtrace(exception)
+    end
+
+    begin
+      MonitorPipelineResults.fail_stalled_uploads!
+    rescue => exception
+      LogUtil.log_err_and_airbrake("Failed to fail stalled uploads: #{exception.message}")
+      LogUtil.log_backtrace(exception)
+    end
+  end
+
+  def self.fail_stalled_uploads!
+    # Delay determined by doubling the time after which an alert is sent to admins.
+    Sample
+      .stalled_uploads(6.hours)
+      .each { |sample| sample.fail_local_upload!("Stalled for more than 6 hours") }
```

Comment:
Could we do the status update in batch using `update_all`? Adding O(N) synchronous operations to result monitor is a scalability no-go.

Reply:
We could. I didn't think it would be an issue because it is in the error path. If we have so many upload failures that the performance of the result monitor is affected, then we have bigger problems 🙂

Reply:
I still think you should. The cycle is already a bottleneck when there many samples and we need every perf gain we can get, especially when it’s so easy.

Reply:
Updated